### ISSUE TITLE
Visual Crossing doesn't connect anymore with newest version

### ISSUE THREAD

**Issue by Olifant1990** [domoticz/domoticz#6176]
Last week I updated to version 2024_7.16253 (running on a fully updated Windows 10 PC).
I got an error with Visual Crossing: "2024-09-22 21:57:45.811 Error: Visual Crossing: Error getting http data!."
So, mark the dot at the end, it brings me to https://github.com/domoticz/domoticz/blob/03baca33331c03c5cfd295ec546a2eca151b17c8/hardware/VisualCrossing.cpp#L133

But... When I run the latest code on my development PC (Windows 11), build with Visual Studio, I do not get this error and all is working fine.
Any ideas where to look for this?
Problem is still present in current beta 2024_7.16265

**Comment by waltervl** [comment]
Seems an internet connection issue on your W10 machine.
What do you get on this link when used on the browser on that W10 machine. replace szLoc  with the contents of your location used in intergration or Domoticz settings.
And change API field
`https://weather.visualcrossing.com/VisualCrossingWebServices/rest/services/timeline/" << szLoc << "?key=" << m_APIKey << "&unitGroup=metric&include=current`

**Comment by Olifant1990** [comment]
Sorry, I forgot to mention that the direct JSON link IS working in a browser.
Same as that version 2024_7.16187 is working, but newer aren't.

**Comment by kiddigital** [comment]
@Olifant1990 , as the connection is using HTTPS (so SSL), there is a good chance that your W10 PC does not has the right SSL library available for Domoticz.

The browser has often a internal SSL library.

**Comment by Olifant1990** [comment]
Then somewhere between version 2024_7.16187 and 2024_7.16253 that must have been changed.
As I stick to 2024_7.16187 for now and it is working for now.

**Comment by Olifant1990** [comment]
Any ideas on how to fix this?

**Comment by gizmocuz** [comment]
The last code change for this hardware class has been made on 9 September 2023... (after this only one change that has no impact)
So maybe there has been a change on your side?

It is also a very small class with not many code lines... Maybe you can have a look at it: VisualCrossing.cpp

**Comment by Olifant1990** [comment]
Well, as said, It is not a problem of VisualCrossing itself. I know the code as I am the writer of it :P
But... Also as said, see pointed line in first post.
if (!HTTPClient::GET(sURL.str(), sResult)) is giving false in later versions, so I saw that error comming from visual crossing.
I went back to 2024_7.16187 and it is still working for now on that version.
If I update, visual crossing is not able to get response via the HTTPClient::Get.
In browser, visual crossing is giving me a correct JSON.

I will dig deeper as far as I am able to do, but I myself have no clue yet :(

Maybe something changed in the CURL lib?

**Comment by gizmocuz** [comment]
Nah that would mean hundred thousend of users have issues with this.
And as HTTPClient::GET is used by many hardware (also hardware I am using), we would have seen many more reports of failures

**Comment by gizmocuz** [comment]
I know they drop the old API and you have to use the new API where you will have to pay above a certain amount of queries (1000?), maybe you could try to create a new account?

**Comment by Olifant1990** [comment]
No, visual crossing is fine.
Domoticz cannot do the GET request on my Win10 64 bit pc.
In the dashboard of visual crossing I see no requests on the newest beta. When reverting back, it is working.

Also when requesting the URL directly, I a request count on the dashboard of Visual Crossing and get a result

**Comment by kiddigital** [comment]
@Olifant1990 , so on this 1 PC it is not working with the latest release build of Domoticz, but is working with a previous version of Domoticz?

Do you have access to an updated Windows 11 machine to test if the same latest release of Domoticz does or does not work on that machine?

This is to test if the problem could be due to underlying OS libraries (suspecting things like SSL, etc).

**Comment by gizmocuz** [comment]
Is it possible to launch another instance of Domoticz.
Only add the visual crossing hardware
Is this working?

I added the hardware, it's been quite some time ago since I used it.
When getting the weather data, I received a HTTP issue as well, I added more logging to this and this is the result:

2024-10-22 07:47:41.811  Error: Visual Crossing: Error getting http data!.
2024-10-22 07:47:43.432  Error: Visual Crossing: You do not have an active plan that allows for an API Key. Please verify your account details.

This was also what I said in the previous message, they changed their rules/API usage.

Indeed, I did not had a plan. I added a plan

I had to press the option "Pricing and Editions" otherwise I could not select the free plan
But when I do so, I quickly see a page to enter an email address , and half a second later it is being redirected to a weather lookup page.
So, I am not able to add the free plan

I can't help you with this one.

Please debug with the latest code (just committed my error log addon)

**Comment by Olifant1990** [comment]
@gizmocuz I wanted to test with the latest beta today, but I see that there is no new build since 16-10 for Windows?
Maybe a problem on the build server?

**Comment by gizmocuz** [comment]
Yes we have issues with the Windows build, there are no new changes to visual crossing made after this date as far as I know
I thought you where able to compile yourself?
I uploaded the latest Windows setup just now

**Comment by Olifant1990** [comment]
@gizmocuz I am indeed able to build it myself, so I know the most recent code is working on windows 11.
But I cannot build and use that on my production server (win 10). So I needed the new build because it contained your more logging in VisualCrossing.cpp.

But

I tried lates build today.
Still failing on my windows 10, your extra log row is telling me: "Bad API Request:Invalid location parameter value".
So it is reaching visual crossing, but with wrong parameters or something.
Maybe an encoding issue somewhere.
The URL build by Visual Crossing, pasted in my browser on that PC is working correctly.
Same location (lat,long) is working on my win 11 PC.

I pushed a new PR with extended log, URL included.
I hope to figure out what the problem is that way.

**Comment by waltervl** [comment]
Is your location (menu setup-settings) in your Domoticz w10 and W11 environments exactly the same (incl dots and commas)?

**Comment by Olifant1990** [comment]
Yes it is.
I copy the database from live to my development and start with the latest code. Then it is working.
But somewhere after 2024_7.16187 it is not working because of that bad location parameter on my windows 10 live PC

**Comment by Olifant1990** [comment]
Still on it!

**Comment by Olifant1990** [comment]
@gizmocuz 

It is not solved by this PR.
But I now know the problem...

I now see a log row " Error: Visual Crossing: Error getting http data for location `xx.xxxxxxx%0Cx.xxxxxxx`!."
(I have hidden my location).

The code in visual crossing in https://github.com/domoticz/domoticz/blob/406d214febd2e4f611b23b53b2a01053d6f56a36/hardware/VisualCrossing.cpp#L127 is encoding the comma wrong...
, is being translated to %0C, while visual crossing expects a %2C (or a comma, without url encode).

Do you know how to fix this?

**Comment by gizmocuz** [comment]
That's odd, as when I enter 5,52 as location, this get's perfectly translated to

5%2C52

And when logged:

Error: VisualCrossing: Error getting http data for location `5%2C52`!

Error: VisualCrossing: You do not have an active plan that allows for an API Key. Please verify your account details

%0C is not correct.

Could you try with 5,52? (maybe the problem is something else)

**Comment by Olifant1990** [comment]
@gizmocuz 

5,52 is also translated to 5%0C52 at my PC... Strange, no clue why.

And can you reopen this issue for me?

**Comment by waltervl** [comment]
Fixed?

**Comment by Olifant1990** [comment]
@waltervl 

No, but it seems to be only a problem on my Domoticz instance.
When I save my database file and start a new instance, all is fine.
Even on all my other PCs when I try it on a new database.
So for now I changed the coördinates to a placename, which also works...
But it is still strange.

**Comment by gizmocuz** [comment]
Okey, Let's close the issue. Feel free to re-open if it is reproducible
